src/source/agc.rs

```diff
@@ -0,0 +1,175 @@
+use super::SeekError;
+use crate::{Sample, Source};
+use std::time::Duration;
+
+/// Constructs an `AutomaticGainControl` object with specified parameters.
+///
+/// # Arguments
+///
+/// * `input` - The input audio source
+/// * `target_level` - The desired output level
+/// * `attack_time` - Time constant for gain adjustment
+/// * `absolute_max_gain` - Maximum allowable gain
+pub fn automatic_gain_control<I>(
+    input: I,
+    target_level: f32,
+    attack_time: f32,
+    absolute_max_gain: f32,
+) -> AutomaticGainControl<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    let sample_rate = input.sample_rate();
+
+    AutomaticGainControl {
+        input,
+        target_level,
+        absolute_max_gain,
+        current_gain: 1.0,
+        attack_coeff: (-1.0 / (attack_time * sample_rate as f32)).exp(),
+        peak_level: 0.0,
+        rms_level: 0.0,
+        rms_window: vec![0.0; 1024],
+        rms_index: 0,
+    }
+}
+
+/// Automatic Gain Control filter for maintaining consistent output levels.
+#[derive(Clone, Debug)]
+pub struct AutomaticGainControl<I> {
+    input: I,
+    target_level: f32,
+    absolute_max_gain: f32,
+    current_gain: f32,
+    attack_coeff: f32,
+    peak_level: f32,
+    rms_level: f32,
+    rms_window: Vec<f32>,
+    rms_index: usize,
+}
+
+impl<I> AutomaticGainControl<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    // Sets a new target output level.
+    #[inline]
+    pub fn set_target_level(&mut self, level: f32) {
+        self.target_level = level;
+    }
+
+    // Add this method to allow changing the attack coefficient
+    pub fn set_attack_coeff(&mut self, attack_time: f32) {
+        let sample_rate = self.input.sample_rate();
+        self.attack_coeff = (-1.0 / (attack_time * sample_rate as f32)).exp();
+    }
+}
+
+impl<I> Iterator for AutomaticGainControl<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    type Item = I::Item;
+
+    #[inline]
+    fn next(&mut self) -> Option<I::Item> {
```

Comment:
a nitpick (Its fine not to change this) but I was wondering how you would feel about extracting some of this code into member functions. To me the first 3 comments almost beg to become a function name.

self.update_peak_level();
self.update_rms_level();
let gain = self.desired_gain();

Reply:
That's a good idea thanks :)

Im working on it now and cleaning it up a bit + better comments